src/inc/til/spsc.h

```diff
@@ -0,0 +1,526 @@
+        template<typename T>
+        struct arc
+        {
+            arc(size_type capacity) noexcept :
+                _data(alloc_raw_memory<T>(size_t(capacity) * sizeof(T))),
+                _capacity(capacity)
+            {
+            }
+
+            ~arc()
+            {
+                auto beg = _consumer.load(std::memory_order_acquire);
+                auto end = _producer.load(std::memory_order_acquire);
+                auto differentRevolution = ((beg ^ end) & revolution_flag) != 0;
+
+                beg &= position_mask;
+                end &= position_mask;
+
+                // The producer position will always be ahead of the consumer, but since we're dealing
+                // with a ring buffer the producer may be wrapped around the end of the buffer.
+                // We thus need to deal with 3 potential cases:
+                // * No valid data.
+                //   If both positions including their revolution bits are identical.
+                // * Valid data in the middle of the ring buffer.
+                //   If _producer > _consumer.
+                // * Valid data at both ends of the ring buffer.
+                //   If the revolution bits differ, even if the positions are otherwise identical,
+                //   which they might be if the channel contains exactly as many values as its capacity.
+                if (end > beg)
+                {
+                    std::destroy(_data + beg, _data + end);
+                }
+                else if (differentRevolution)
+                {
+                    std::destroy(_data, _data + end);
+                    std::destroy(_data + beg, _data + _capacity);
+                }
+
+                free_raw_memory(_data);
+            }
+
+            void drop_producer()
+            {
+                drop(_producer);
+            }
+
+            void drop_consumer()
+            {
+                drop(_consumer);
+            }
+
+            acquisition producer_acquire(size_type slots) noexcept
+            {
+                return acquire(_producer, _consumer, revolution_flag, slots);
+            }
+
+            void producer_release(acquisition acquisition) noexcept
+            {
+                release(_producer, acquisition);
+            }
+
+            acquisition consumer_acquire(size_type slots) noexcept
+            {
+                return acquire(_consumer, _producer, 0, slots);
+            }
+
+            void consumer_release(acquisition acquisition) noexcept
+            {
+                release(_consumer, acquisition);
+            }
+
+            T* data() const noexcept
+            {
+                return _data;
+            }
+
+        private:
+            void drop(atomic_size_type& mine)
+            {
+                const auto myPos = mine.load(std::memory_order_relaxed);
+                mine.store(myPos | drop_flag, std::memory_order_release);

```

Comment:
As with the other comment, this needs to be release, because only that way we can ensure that the other thread actually sees any change in the atomic's value.